pkg/controller/keyvault/gateway_secret_provider_class.go

```diff
@@ -0,0 +1,231 @@
+		return ctrl.Result{}, nil
+	}
+
+	// check its TLS options - needs to have both cert uri and either serviceaccount name or clientid
+	for index, listener := range gwObj.Spec.Listeners {
+		spc := &secv1.SecretProviderClass{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "secrets-store.csi.x-k8s.io/v1",
+				Kind:       "SecretProviderClass",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      generateGwListenerCertName(gwObj.Name, listener.Name),
+				Namespace: req.Namespace,
+				Labels:    manifests.GetTopLevelLabels(),
+				OwnerReferences: []metav1.OwnerReference{{
+					APIVersion: gwObj.APIVersion,
+					Controller: util.ToPtr(true),
+					Kind:       gwObj.Kind,
+					Name:       gwObj.Name,
+					UID:        gwObj.UID,
+				}},
+			},
+		}
+		logger = logger.WithValues("spc", spc.Name)
+
+		if listenerIsKvEnabled(listener) {
+			var clientId string
+			clientId, err = retrieveClientIdForListener(ctx, g.client, req.Namespace, listener.TLS.Options)
+			if err != nil {
+				var userErr userError
+				if errors.As(err, &userErr) {
+					logger.Info(fmt.Sprintf("failed to fetch clientId for SPC for listener %s due to user error: %s, sending warning event", listener.Name, userErr.userMessage))
+					g.events.Eventf(gwObj, corev1.EventTypeWarning, "InvalidInput", "invalid TLS configuration: %s", userErr.userMessage)
+					return ctrl.Result{}, nil
+				}
+				logger.Error(err, fmt.Sprintf("failed to fetch clientId for listener %s: %s", listener.Name, err.Error()))
+				return ctrl.Result{}, fmt.Errorf("fetching clientId for listener: %w", err)
+			}
+
+			// otherwise it's active + valid - build SPC
+			certUri := string(listener.TLS.Options[certUriTLSOption])
+
+			logger.Info("building spc for listener and upserting")
+			spcConf := spcConfig{
+				ClientId:        clientId,
+				TenantId:        g.config.TenantID,
+				KeyvaultCertUri: certUri,
+				Name:            generateGwListenerCertName(gwObj.Name, listener.Name),
+			}
+			err = buildSPC(spc, spcConf)
+			if err != nil {
+				var userErr userError
+				if errors.As(err, &userErr) {
+					logger.Info("failed to build SecretProviderClass from user error: %s, sending warning event", userErr.userMessage)
+					g.events.Eventf(gwObj, corev1.EventTypeWarning, "InvalidInput", "invalid TLS configuration: %s", userErr.userMessage)
+					return ctrl.Result{}, nil
+				}
+				logger.Error(err, fmt.Sprintf("building SPC for listener %s: %s", listener.Name, err.Error()))
+				return ctrl.Result{}, fmt.Errorf("building spc: %w", err)
+			}
+
+			logger.Info(fmt.Sprintf("reconciling SecretProviderClass %s for listener %s", spc.Name, listener.Name))
+			if err = util.Upsert(ctx, g.client, spc); err != nil {
+				fullErr := fmt.Errorf("failed to reconcile SecretProviderClass %s: %w", req.Name, err)
+				logger.Error(err, fullErr.Error())
+				g.events.Event(gwObj, corev1.EventTypeWarning, "FailedUpdateOrCreateSPC", fullErr.Error())
+				return ctrl.Result{}, fullErr
+			}
```

Comment:
consider creating a helper function for repeated util.Upsert calls to improve code reusability!

Reply:
I might be missing something but what're you referring to? We only call upsert once here for the secretproviderclass, and then a regular client update below (see the thread)